internal/host/host.go

```diff
@@ -207,6 +207,47 @@ func (m *Manager) HandleInstallationFailure(ctx context.Context, h *models.Host)
 	return err
 }
 
+// populateDisksEligibility updates an inventory json string by updating the eligibility
+// struct of each disk in the inventory with service-side checks for disk eligibility, in
+// addition to agent-side checks that have already been performed. The reason that some
+// checks are performed by the agent (and not the service) is because the agent has data
+// that is not available in the service.
+func (m *Manager) populateDisksEligibility(inventoryString string) (string, error) {
+	var inventory models.Inventory
+	if err := json.Unmarshal([]byte(inventoryString), &inventory); err != nil {
+		return "", err
+	}
+
+	for _, disk := range inventory.Disks {
+		if !disk.InstallationEligibility.Eligible && len(disk.InstallationEligibility.NotEligibleReasons) == 0 {
+			// This is a sign that the agent is using an old version that does not fill-in the eligibility struct,
+			// for backwards compatibility, pretend that the agent has decided that this disk is eligible
+			disk.InstallationEligibility.Eligible = true
+			disk.InstallationEligibility.NotEligibleReasons = make([]string, 0)
+		} else if disk.InstallationEligibility.Eligible && len(disk.InstallationEligibility.NotEligibleReasons) != 0 {
+			// The agent decided that it's eligible. Clean up any reasons that might somehow have ended up here
+			m.log.Warnf("Agent reported that disk is eligible but the reasons array was not empty: %v",
+				disk.InstallationEligibility.NotEligibleReasons)
+			disk.InstallationEligibility.NotEligibleReasons = make([]string, 0)
+		}
+
+		eligible, newNotEligibleReasons := m.hwValidator.DiskIsEligible(disk)
+
+		// Append to the existing reasons already filled in by the agent
+		disk.InstallationEligibility.NotEligibleReasons = append(disk.InstallationEligibility.NotEligibleReasons, newNotEligibleReasons...)
+
+		// Combine the final decision of both the agent and the service
+		disk.InstallationEligibility.Eligible = disk.InstallationEligibility.Eligible && eligible
```

Comment:
A bit of a nit, but this feels strange ... Since the definition of the eligible boolean in the validator is just `len(NotEligibleReasons) > 0` why are these separate? It feels like we should just be passing around the reasons and use `len(NotEligibleReasons) > 0` here to set `Eligible`.

Reply:
If we do that, how will we be able to distinguish between "is eligible" and "eligibility undetermined" (older agents)?

At first I wanted to add a "is_set" flag to this struct but realized `false` with empty reasons can be used to detect whether it was set or not (this was before I knew it would actually use a pointer for the struct, then I just used the null pointer as an indication. Then we discussed it and changed it to `is-nullable: false`).

Reply:
Oh, I'm not suggesting removing the "column", I was just questioning the necessity of multiple return values from `DiskIsEligible`, but really I wouldn't change it now. Maybe it's just that it looked like a conditional was missing here, but the append should be a noop if we get `eligible == true` so not a big deal. I'll just resolve this one.